base_layer/core/src/proof_of_work/diff_adj_manager/diff_adj_storage.rs

```diff
@@ -314,43 +328,22 @@ impl DiffAdjStorage {
     ) -> Result<(), DiffAdjManagerError>
     {
         self.reset();
-        let mut monero_diff_list = Vec::<(EpochTime, Difficulty)>::with_capacity(self.difficulty_block_window as usize);
-        let mut blake_diff_list = Vec::<(EpochTime, Difficulty)>::with_capacity(self.difficulty_block_window as usize);
-        for height in (0..=height_of_longest_chain).rev() {
+        for height in 0..=height_of_longest_chain {
             let header = fetch_header(db, height)?;
             // keep MEDIAN_TIMESTAMP_COUNT blocks for median timestamp
             if self.timestamps.len() < self.median_timestamp_count {
                 self.timestamps.push_front(header.timestamp);
             }
             match header.pow.pow_algo {
                 PowAlgorithm::Monero => {
-                    if (monero_diff_list.len() as u64) < self.difficulty_block_window {
-                        monero_diff_list.push((
-                            header.timestamp,
-                            header.pow.accumulated_monero_difficulty + header.achieved_difficulty(),
-                        ));
-                    }
+                    let target_difficulty = self.monero_lwma.get_difficulty();
+                    self.monero_lwma.add(header.timestamp, target_difficulty)?;
                 },
                 PowAlgorithm::Blake => {
-                    if (blake_diff_list.len() as u64) < self.difficulty_block_window {
-                        blake_diff_list.push((
-                            header.timestamp,
-                            header.pow.accumulated_blake_difficulty + header.achieved_difficulty(),
-                        ));
-                    }
+                    let target_difficulty = self.blake_lwma.get_difficulty();
+                    self.blake_lwma.add(header.timestamp, target_difficulty)?;
                 },
             }
-            if ((monero_diff_list.len() as u64) >= self.difficulty_block_window) &&
-                ((blake_diff_list.len() as u64) >= self.difficulty_block_window)
-            {
-                break;
```

Comment:
This should break or you will query every single header in your blockchain.

Reply:
Unfortunately you have to at the moment because we don't store the target difficulty anyway. I'll put in a TODO to store the target difficulty